blockindex/liquidstaking_indexer.go

```diff
@@ -0,0 +1,687 @@
+	s.mutex.RLock()
+	defer s.mutex.RUnlock()
+
+	vbs := []*ContractStakingBucket{}
+	for id, bi := range s.cache.getAllBucketInfo() {
+		bt := s.cache.mustGetBucketType(bi.TypeIndex)
+		vb, err := s.assembleContractStakingBucket(id, bi, bt)
+		if err != nil {
+			return nil, err
+		}
+		vbs = append(vbs, vb)
+	}
+	return vbs, nil
+}
+
+// Bucket returns the bucket
+func (s *liquidStakingIndexer) Bucket(id uint64) (*ContractStakingBucket, error) {
+	s.mutex.RLock()
+	defer s.mutex.RUnlock()
+
+	return s.generateBucket(id)
+}
+
+// BucketsByIndices returns the buckets by indices
+func (s *liquidStakingIndexer) BucketsByIndices(indices []uint64) ([]*ContractStakingBucket, error) {
+	s.mutex.RLock()
+	defer s.mutex.RUnlock()
+
+	vbs := make([]*ContractStakingBucket, 0, len(indices))
+	for _, id := range indices {
+		vb, err := s.generateBucket(id)
+		if err != nil {
+			return nil, err
+		}
+		vbs = append(vbs, vb)
+	}
+	return vbs, nil
+}
+
+func (s *liquidStakingIndexer) TotalBucketCount() uint64 {
+	return s.cache.getTotalBucketCount()
+}
+
+func (s *liquidStakingIndexer) ActiveBucketTypes() (map[uint64]*ContractStakingBucketType, error) {
+	return s.cache.getActiveBucketType(), nil
+}
+
+func (s *liquidStakingIndexer) generateBucket(id uint64) (*ContractStakingBucket, error) {
+	bi, ok := s.cache.getBucketInfo(id)
+	if !ok {
+		return nil, errors.Wrapf(ErrBucketInfoNotExist, "id %d", id)
+	}
+	bt := s.cache.mustGetBucketType(bi.TypeIndex)
+	return s.assembleContractStakingBucket(id, bi, bt)
+}
+
+func (s *liquidStakingIndexer) handleEvent(ctx context.Context, dirty *liquidStakingDirty, blk *block.Block, log *action.Log) error {
+	// get event abi
+	abiEvent, err := _liquidStakingInterface.EventByID(common.Hash(log.Topics[0]))
+	if err != nil {
+		return errors.Wrapf(err, "get event abi from topic %v failed", log.Topics[0])
+	}
+
+	// unpack event data
+	event, err := unpackEventParam(abiEvent, log)
+	if err != nil {
+		return err
+	}
+
+	// handle different kinds of event
+	switch abiEvent.Name {
+	case "BucketTypeActivated":
+		return dirty.handleBucketTypeActivatedEvent(event, blk.Height())
+	case "BucketTypeDeactivated":
+		return dirty.handleBucketTypeDeactivatedEvent(event, blk.Height())
+	case "Staked":
+		return dirty.handleStakedEvent(event, blk.Height())
+	case "Locked":
+		return dirty.handleLockedEvent(event)
+	case "Unlocked":
+		return dirty.handleUnlockedEvent(event, blk.Height())
+	case "Unstaked":
+		return dirty.handleUnstakedEvent(event, blk.Height())
+	case "Merged":
+		return dirty.handleMergedEvent(event)
+	case "DurationExtended":
+		return dirty.handleDurationExtendedEvent(event)
+	case "AmountIncreased":
+		return dirty.handleAmountIncreasedEvent(event)
+	case "DelegateChanged":
+		return dirty.handleDelegateChangedEvent(event)
+	case "Withdrawal":
+		return dirty.handleWithdrawalEvent(event)
+	case "Transfer":
+		return dirty.handleTransferEvent(event)
+	default:
+		return nil
```

Comment:
should inconsistency be ignored here?

Reply:
good question, events that do not require handling should be clearly labeled and `return nil`, while unexpected events should `return error`.